### PR TITLE
Replace ProcessModule.FileName with Environment.ProcessPath and remove PSUtils.GetMainModule

### LINKED ISSUE #13790: Use Environment.ProcessPath API
Replace Process.GetCurrentProcess().MainModule.FileName with new API Environment.ProcessPath
(We use this only in test tools.)

https://github.com/dotnet/runtime/issues/42948

We will get the API only after we move to .Net 6.0 Preview1.

### PR DESCRIPTION
**Pull request by Fs00**
# PR Summary

This PR replaces all occurrences of `PsUtils.GetMainModule(currentProcess).FileName` with the new `Environment.ProcessPath` API as suggested in https://github.com/PowerShell/PowerShell/pull/14332#issuecomment-739549844.
Furthermore, all the other places in which a ProcessModule needs to be retrieved have been changed to use `Process.MainModule` .NET API directly instead of going through `PSUtils.GetMainModule` (which has been removed).

## PR Context

In the comment linked above, @iSazonov reported that calling `PsUtils.GetMainModule(currentProcess).FileName` was particularly expensive (27ms). The new `Environment.ProcessPath` API has been implemented to tackle this problem and there is even an official .NET analyzer to recommend switching to it (https://github.com/dotnet/roslyn-analyzers/pull/4909).
While making that change, I noticed that `PSUtils.GetMainModule` wrapped `Process.MainModule` in a retry loop to workaround a .NET Framework 2.0 (!) BCL issue, therefore it shouldn't be needed at all nowadays.

Fixes #13790

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/security/SecuritySupport.cs
Comment (Fs00) [inline src/System.Management.Automation/security/SecuritySupport.cs:1355]: In a further PR I will remove the need to use `Process.MainModule.FileVersionInfo` here to avoid loading System.Diagnostics.FileVersionInfo.dll assembly on startup on Windows.

File: src/System.Management.Automation/security/SecuritySupport.cs
Comment (iSazonov) [inline src/System.Management.Automation/security/SecuritySupport.cs:251]: We need a conclusion from @PaulHigin and @TravisEz13 

Looking [the API implementation](https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L193-L203) I believe it is safe change - main module has already loaded.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:665]: It is not critical in the context and we can accept the change.

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs
Comment (Fs00) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:685]: I agree that it looks like a bug, but I didn't touch it because otherwise I would have changed existing behavior.
If you are OK to fix it in this PR, then it's fine for me.

File: src/System.Management.Automation/security/SecuritySupport.cs
Comment (Fs00) [reply src/System.Management.Automation/security/SecuritySupport.cs:251]: What do you mean by "main module has already loaded"?
We cannot use Environment.ProcessPath here because `currentProcess` refers to one of the parent processes of PowerShell (that variable is wrongly named).

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:685]: I believe we never want NRE in the context.

Comment (Fs00) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:685]: Sure, my doubt was about doing that change in another PR rather than in this one.
I agree that we should do it while we're at it.

File: src/System.Management.Automation/security/SecuritySupport.cs
Comment (iSazonov) [reply src/System.Management.Automation/security/SecuritySupport.cs:251]: To get the main module the API enumerates all modules. Docs say the enumeration can be interrupted if an assembly is loading in the time or main module is still initialized. 
I guess it is safe to ignore this in the context. In any case, this is not a reliable and predictable solution since we cannot predict when an assembly will load.

Comment (Fs00) [reply src/System.Management.Automation/security/SecuritySupport.cs:251]: In any case, using Process.MainModule directly does not affect reliability since the .NET BCL has a [built-in retry logic for the edge case you mentioned](https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs#L204-L224).
It seems that `PsUtils.GetMainModule` was made before the retry logic was implemented in the .NET BCL.

Comment (TravisEz13) [reply src/System.Management.Automation/security/SecuritySupport.cs:251]: MainModule could be null if the process is not loaded:  
https://github.com/dotnet/runtime/blob/6d238511167bdbf8cd16103e84b6befb45c0e301/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L200
The question is if that is relavant.

Comment (TravisEz13) [reply src/System.Management.Automation/security/SecuritySupport.cs:251]: Reviewing the code, it seems to be handled, where we aren't using currentProcess.  I think we are okay.

**Review by TravisEz13 — COMMENTED** [review]
I looks okay to me.

**Comment by Fs00** [comment]
I see that this PR has been already approved, is anything else needed to get this merged?

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
We should have approval from one of the code owners either @adityapatwardhan  or @daxian-dbw

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@Fs00 Thank you for your contribution!

**Comment by Fs00** [comment]
You're welcome!👍🏻

**Pull request merged**
